Add jumps and conditional jumps

a2aaron opened the pull request.

a2aaron committed:
Add jumps and conditional jumps

porglezomp commented on src/lib.rs:
```diff
@@ -18,6 +18,10 @@ pub enum Instr {
     /// a = b % c
     Rem(Addr, Addr, Addr),
+    /// Jumps program execution by n instructions
+    Jump(i8),
```
We could use a larger value (`i16`?) for our unconditional jump, since it wouldn't make the instruction size larger.

porglezomp commented on src/lib.rs:
```diff
@@ -18,6 +18,10 @@ pub enum Instr {
     Rem(Addr, Addr, Addr),
+    /// Jumps program execution by n instructions
+    Jump(i8),
+    /// Jumps program execution by n instructions if a is non zero, else it jumps by m instructions
```
We have a proper boolean type (the `B` value), so we shouldn't be working in terms of “nonzero.”

porglezomp commented on src/lib.rs:
```diff
@@ -112,6 +116,18 @@ impl Program {
+                    continue;
+                }
+                CondJump(a, b, c) => {
+                    if locals[a as usize] != I(0) {
```
We should consider `B(false)` to be false.
If we want to be lenient, we should also consider both `I(0)` and `F(0.0)` as well.
If we don't want to be as lenient (I'm leaning towards this), we might want to throw an error if we get a non-boolean value here.

porglezomp commented on src/lib.rs:
```diff
@@ -112,6 +116,18 @@ impl Program {
                 }
+                Jump(a) => {
+                    iptr += a as isize;
+                    continue;
```
Ah good you used the `continue` method, I missed this on my first pass and did it this way in the fixes I just wrote. :)

a2aaron committed:
Make CondJump check for type

If you attempt to use CondJump on a non-boolean type, the attempt will now return an Err

porglezomp commented on src/lib.rs:
```diff
@@ -112,6 +117,22 @@ impl Program {
+                    continue;
+                }
+                CondJump(a, b, c) => {
+                    if let Val::B(x) = locals[a as usize] {
```
You could probably do a `match` and match `Val::B(true)`, `Val::B(false)`, and `_`, if that looks nicer. Like:
```rust
match locals[a as usize] {
    B(true) => …,
    B(false) => …,
    _ => return …,
}
continue;
```

porglezomp replied:
Also you don't need `Val::B` since we've done `use Val::*;`

a2aaron committed:
Clean up match statement for CondJump

a2aaron committed:
Make backwards jumps work.

porglezomp commented on src/lib.rs:
```diff
@@ -130,12 +135,32 @@ impl Program {
     }
 }
 
+fn sum(a: usize, b: isize) -> usize {
```
I think a name like `offset_index` or something similar would be better

a2aaron committed:
Add Eq and Neq

a2aaron committed:
Add comparative operators.

porglezomp commented on src/lib.rs:
```diff
@@ -83,6 +101,12 @@ impl Program {
                 &Mul(a, b, c) => locals[a as usize] = (&locals[b as usize] * &locals[c as usize])?,
                 &Div(a, b, c) => locals[a as usize] = (&locals[b as usize] / &locals[c as usize])?,
                 &Rem(a, b, c) => locals[a as usize] = (&locals[b as usize] % &locals[c as usize])?,
+                &Eq(a, b, c) => locals[a as usize] = B(&locals[b as usize] == &locals[c as usize]),
```
The derived `PartialEq` impl will consider `I(0)` and `F(0.0)` different. This is probably fine, but it means that we need a numeric cast operator. (Perhaps just make an issue?)

porglezomp commented on src/lib.rs:
```diff
@@ -83,6 +101,12 @@ impl Program {
                 &Div(a, b, c) => locals[a as usize] = (&locals[b as usize] / &locals[c as usize])?,
                 &Rem(a, b, c) => locals[a as usize] = (&locals[b as usize] % &locals[c as usize])?,
+                &Eq(a, b, c) => locals[a as usize] = B(&locals[b as usize] == &locals[c as usize]),
+                &Neq(a, b, c) => locals[a as usize] = B(&locals[b as usize] != &locals[c as usize]),
```
Since these traits act by reference, does using `locals[b as usize] != locals[c as usize]` work, or do we get "moved out of indexed context"?

a2aaron committed:
Make a macro for testing programs

a2aaron committed:
Macroify the other tests

porglezomp commented on src/lib.rs:
```diff
@@ -196,82 +240,189 @@ impl<'a> Rem for &'a Val {
+            code: [$($instrs:expr),* $(,)*],
+            consts: [$($consts:expr),* $(,)*],
+            local_count: $count:expr,
+        },)*
```
change `},)*` to just `})*`, and the tests to match

porglezomp reviewed (approved):
LGTM
